src/ioctls/vcpu.rs

```diff
@@ -1220,11 +1220,12 @@ impl VcpuFd {
     ///
     /// * `reg_id` - ID of the register.
     #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
-    pub fn get_one_reg(&self, reg_id: u64) -> Result<u64> {
-        let mut reg_value = 0;
+    pub fn get_one_reg<T: Sized>(&self, reg_id: u64) -> Result<T> {
+        // SAFETY: Zeroing the value for the kernel to write to
+        let mut reg_value: T = unsafe { std::mem::zeroed() };
```

Comment:
What happens if the user is passing a reference type as T? I don't think this will be safe anymore.

Reply:
You get a runtime error which is easy to catch. https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=66745c63561329a210caf1cc09936fa8

Reply:
Then it means we should mark the function unsafe and update its documentation.

We should also add some tests exercising the new API, e.g. call the functions with u128 instead of u64.

Reply:
I'm not sure if we should just add a `Panics` section instead of marking the whole function unsafe, but according to [mem::zeroed()](https://doc.rust-lang.org/std/mem/fn.zeroed.html) documentation, we get undefined behavior when we use references (it doesn't explicitly say that we panic), so I think it would be safer to mark the function as unsafe.

Reply:
I just attempted to write a negative trait bound to rule out `Deref` but that's not allowed in Rust yet.

Marking it unsafe is of course fine. Alternative, the return type can be made `u128` so that it stays a safe function. And then perhaps have a few checks in the function to own accept below 128-bit register types.

The original implementation is already unsound (safe function can violate memory safety) so any attempt is one step towards the right direction.

Reply:
The alternative is to use `MaybeUninit`:

```rust
    pub fn get_one_reg<T: Sized + Copy>(&self, reg_id: u64) -> Result<T> {
        let mut reg_value = std::mem::MaybeUninit::<T>::uninit();
        let mut onereg = kvm_one_reg {
            id: reg_id,
            // SAFETY: Passing a temporary value on the stack that is sized
            addr: unsafe { reg_value.as_mut_ptr() as u64 },
        };

        let ret = unsafe { ioctl_with_mut_ref(self, KVM_GET_ONE_REG(), &mut onereg) };
        if ret < 0 {
            return Err(errno::Error::last());
        }

        // SAFETY: Value has been initialised by the kernel
        Ok(unsafe { reg_value.assume_init() })
    }
```

Reply:
Also, we could check the size of the register (as it's embedded in `reg_id`) matches the size of the datatype.

Reply:
We can limit the types by having our own marker trait.

```
trait Unsigned {}

impl Unsigned for u16 {}
impl Unsigned for u32 {}
impl Unsigned for u64 {}
impl Unsigned for u128 {}

fn f<T: Sized + Unsigned>() -> Result<T, ()> {
    let x = std::mem::MaybeUninit::<T>::uninit();

    Err(())
}

fn main() {
    let _ = f::<u32>();
    let _ = f::<&u32>(); // Rejected
}
```

Reply:
@liuw Nice solution. I've always been disappointed/surprised that there isn't a trait for the basic primitive number types. I've updated this PR with your idea. WDYT?

Reply:
Fine by me of course.

Reply:
Ill have a better look at the solution tomorrow, but just wondering, if we limit T to ValidRegisterType, isn't it safe to go back to the `mem::zeroed()` solution?

Reply:
Also if the trait is public, then the user can still implement it for unsafe types, or am I missing something?

Reply:
@lauralt I think `MaybeUninit` is actually better - this is what it was designed for. Using `std::mem::zeroed` is also wasteful as we don't need to zero the memory being handed to the kernel as it will overwrite it anyway.

I would love it not to be a public trait but it's required to be as it is exposed in the public interface. If all this is too much faff then we can go with the u128 approach.

Reply:
We would probably need to mark the trait as unsafe.. It might be better to use an enum instead, but yeah it becomes overly complicated in my opinion anyway. I think I'm in favor of going with u128, since we are breaking the API either way, and will need turbofish code with the current proposal, as opposed to casting from u64 to u128.

Reply:
I missed this yesterday. Is the decision to go with u128?

Cloud Hypervisor is broken right now with a memory safety issue on ARM, so I would like to get this fixed before the next release, otherwise the release will need to declare ARM save / restore is broken.

Reply:
The problem with the `u128` approach is that it assumes that there won't be `u256` size registers in the near future. The KVM header defines register sizes up to 2048.

Reply:
My preference is still using generic. But then if there is no clear way forward, I think we can do the `u128` change to fix this issue at hand and ponder for a longer period of time whether using generic makes sense. I just cannot unsee that bug once I saw it. :-)

Reply:
Do you have a reference to the possible register values? I think we can go ahead with generics, but we need to make sure to not introduce unsafe behavior.

Reply:
The exhaustive list only exists in KVM's header.

Not sure what you want to do with it. If you want to check all possible registers, that list will change in the future if more registers are made available.

Reply:
I want to look at the current registers. And potentially add a link as a comment to the code as well so people know where to find them. Can you paste a link to the lines in the KVM header that you are referring?

Reply:
I was referring to `arch/arm64/include/uapi/asm/kvm.h`. Other relevant files are `arch/arm64/kvm/guest.c` and `sys_regs.c`.

Reply:
Here I see that the registers are defined as maximum 64 bytes: https://elixir.bootlin.com/linux/latest/source/arch/arm64/include/uapi/asm/kvm.h#L127. Can you use elixir an paste a link please? I do not have so much experience with arm and I think it would help with the review.

Reply:
@andreeaflorescu It's the `fp_regs.vregs` that are u128s: https://elixir.bootlin.com/linux/latest/source/arch/arm64/kvm/guest.c#L92

Reply:
Thanks @rbradford. That's what I was looking for!

Reply:
Anything more to discuss? @andreeaflorescu what's your opinion here?

Reply:
We can use both the trait or u128, I don't have any strong preference. For the trait I think we need to mark it as unsafe as we did for BytedValue in vm-memory. Also, to avoid undefined behavior we need to make sure to zero out the bytes that are not set by KVM.